Fastfile

```diff
@@ -1,175 +1,192 @@
+end
+
+def _build_ipa()
+  update_app_identifier(xcodeproj: ENV['FL_UPDATE_PLIST_PROJECT_PATH'],
+                       plist_path: ENV['FL_UPDATE_PLIST_PATH'],
+                   app_identifier: ENV['FL_UPDATE_PLIST_APP_IDENTIFIER'] )
+  update_info_plist
+  _build_with_gym()
+end
+
+def _build_with_gym()
+  install_provisioning_profile
+  _update_team_id_if_necessary
+  provisioning_profile_name = ENV['TAB_PROVISIONING_PROFILE']
+  # Default export method is enterprise since this is the most commonly used
+  export_method = "enterprise"
+  xcconfig_filename = Dir.pwd + "/TAB.release.xcconfig"
+  if ENV['TAB_EXPORT_METHOD'] != nil
+    export_method = ENV['TAB_EXPORT_METHOD']
+  else
+    UI.message("Fallbacking to enterprise export_method sinceTAB_EXPORT_METHOD is not defined")
+  end
+  if File.file?("Provfile")
+    _parse_provision_file()
+    gym(export_method: export_method, xcconfig: xcconfig_filename)
+  elsif provisioning_profile_name != nil
+    File.write(xcconfig_filename, "PROVISIONING_PROFILE_SPECIFIER = #{provisioning_profile_name}\n")
+    gym(export_method: export_method, xcconfig: xcconfig_filename)
+  else
+    gym(export_method: export_method)
+  end
+end
+
+def _update_team_id_if_necessary()
+  if !ENV['FL_PROJECT_SIGNING_PROJECT_PATH'].to_s.strip.empty? && !ENV['FL_PROJECT_TEAM_ID'].to_s.strip.empty?
+    update_project_team
+  end
+end
+
+def _parse_provision_file()
+  sh 'RUBYSCRIPT="$(echo \'def target(targetName, &doBlock)
+    profileName = doBlock.call
+    print targetName + "_" + "PROVISIONING_PROFILE = " + profileName + "\n"
+  end
+  \' & cat ProvFile)"
+  echo "${RUBYSCRIPT}" | ruby > TAB.release.xcconfig
+  echo \'PROVISIONING_PROFILE_SPECIFIER=$($(TARGET_NAME)_PROVISIONING_PROFILE)\' >> TAB.release.xcconfig'
+end
+
+def _upload_to_hockey()
+  custom_notes = ENV['TAB_HOCKEY_RELEASE_NOTES'] || ""
+  notes = custom_notes == "" ? _create_change_log() : custom_notes
+  hockey(notes_type: "0", notes: notes)
+end
+
+def _create_change_log()
+  cmd = "git log --after={1.day.ago} --pretty=format:'%an%x09%h%x09%cd%x09%s' --date=relative"
+  output = `#{cmd}`
+  return (output.length == 0) ? "No Changes" : output
+end
+
+def _upload_to_test_flight()
+  pilot(username: ENV["ITUNES_CONNECT_USERNAME"],
+    team_id: ENV["ITUNES_CONNECT_TEAM_ID"],
+    itc_provider: ENV["ITUNES_CONNECT_PROVIDER"],
+    skip_waiting_for_build_processing: true,
+    skip_submission: true)
+end
+
+def _notify_slack()
+  if ENV['FL_SLACK_CHANNEL'].to_s.strip.empty?
+    return
+  end
+  hockey_download_url = lane_context[SharedValues::HOCKEY_DOWNLOAD_LINK]
+  if hockey_download_url != nil
+    new_build_message = "A new build is available on <" + hockey_download_url + "|hockey>"
+    slack(message: new_build_message)
+  else
+    slack()
+  end
+end
+
+# --------- After all lanes have run --------- #
+
+after_all do |lane|
+  _notify_slack()
+end
+
+# --------- Error handling --------- #
+
+error do |lane, exception|
```

Comment:
I would argue that this and the `after_all` should sit about the "Custom functions" section. Run the risk of missing them all the way down here.

Reply:
Good shout, will update

Reply:
Updated